posts/2021/09/array-libraries-interoperability.md

```diff
@@ -0,0 +1,578 @@
+
+    return xp
+
+
+def csd(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None, nfft=None,
+		detrend='constant', return_onesided=True, scaling='density', axis=-1,
+		average='mean'):
+    # Array-API
+    xp = get_namespace(x, y)
+
+    # Library Independent Code
+    # Call Array API specified methods
+    some_internal_calculation = xp.some_func(x)
+    ...
+```
+
+This should be possible with the Array API, for the end-user in a
+future release of SciPy and PyTorch.
+
+We made SciPy plus PyTorch work for this very much nontrivial use case,
+which demonstrates feasibility and power of the approach. There are of
+course some issues we ran into, namely
+
+- Lack of Array API standards on complex numbers in the upcoming 2021 version
+  `1.0` of the specification compelled us to special case such instances within the
+  SciPy codebase for PyTorch and NumPy separately.
+- In the current state, the lack of complete Array API compliance in PyTorch
+  is another small issue. One example of PyTorch diverging from NumPy and
+  Array API is `.numel()` vs `.size()`.
+
+All of this is of course addressable with plans to add support for the
+complex numbers module in the second release of Array API Specification.
+We can expect updates and added submodules in a future version of the spec,
+to be released next year.
+
+I'd encourage you to read more details about the
+[demo](https://anirudhdagar.ml/array-api-demo/intro.html) on the tutorial
+webpage itself.
+
+PyTorch developers have been working hard to improve PyTorch's Array API compliance
+fixing divergent behaviour. During the development of this prototype demo,
+I was able to identify some gaps in the [current state of Array API in PyTorch](https://github.com/pytorch/pytorch/labels/module%3A%20python%20array%20api)
+and started my journey as a PyTorch contributor.
+
+
+### Contributing to PyTorch
+
+Interestingly in the process and effort to make PyTorch
+more [Array API](https://data-apis.org/array-api/latest/) compatible,
+I started contributing to PyTorch by raising PRs and relevant Issues.
+My first PyTorch PR [#62560](https://github.com/pytorch/pytorch/pull/62560)
+started with adding an alias `torch.concat` for `torch.cat` which is
+how concatenation is defined in the [array api spec](https://data-apis.org/array-api/latest/API_specification/manipulation_functions.html?highlight=concat#concat-arrays-axis-0)
+and ended with a bunch of other fixes related to concatenation in PyTorch.
+Later, I worked on improving compatibility for Array API to the `torch.linalg`
+module. See [pytorch/pytorch#63285](https://github.com/pytorch/pytorch/pull/63285)
+and [pytorch/pytorch#63227](https://github.com/pytorch/pytorch/pull/63227).
+
+As mentioned earlier, you can track the progress on PyTorch's github repo with
+the label
+["module: python array api"](https://github.com/pytorch/pytorch/labels/module%3A%20python%20array%20api) to check out other interesting developments.
+
+### What's Missing?
```

Comment:
```suggestion
### What's missing?
```

Reply:
I personally like headings to be started with caps unless it is a very long heading. Not sure if there is a convention. If everyone feels that we need to be dramatically correct with caps at the start of the heading only, I can use these suggestions.

Not too rigid on this, just a matter of personal preference. :P